modules/statemachine/statemachinemodule_lua.inl

```diff
@@ -0,0 +1,195 @@
+* The above copyright notice and this permission notice shall be included in all copies *
+* or substantial portions of the Software.                                              *
+*                                                                                       *
+* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,   *
+* INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A         *
+* PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT    *
+* HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF  *
+* CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE  *
+* OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                                         *
+****************************************************************************************/
+
+#include <modules/statemachine/statemachinemodule.h>
+#include <openspace/engine/globals.h>
+#include <openspace/engine/moduleengine.h>
+#include <openspace/scripting/scriptengine.h>
+#include <ghoul/logging/logmanager.h>
+#include <ghoul/misc/misc.h>
+#include <optional>
+
+namespace openspace::luascriptfunctions {
+
+int createStateMachine(lua_State* L) {
+    const int nArguments = ghoul::lua::checkArgumentsAndThrow(
+        L,
+        { 2, 3 },
+        "lua::createStateMachine"
+    );
+
+    // If three arguments, a start state was included
+    std::optional<std::string> startState = std::nullopt;
+    if (nArguments > 2) {
+        startState = ghoul::lua::value<std::string>(L, 3, ghoul::lua::PopValue::Yes);
+    }
+
+    // Last dictionary is on top of the stack
+    ghoul::Dictionary transitions;
+    try {
+        ghoul::lua::luaDictionaryFromState(L, transitions);
+    }
+    catch (const ghoul::lua::LuaFormatException& e) {
+        LERRORC("createStateMachine", e.what());
+        return 0;
+    }
+
+    // Pop, so that first dictionary is on top and can be read
+    lua_pop(L, 1);
+    ghoul::Dictionary states;
+    try {
+        ghoul::lua::luaDictionaryFromState(L, states);
+    }
+    catch (const ghoul::lua::LuaFormatException& e) {
+        LERRORC("createStateMachine", e.what());
+        return 0;
+    }
+
+    StateMachineModule* module = global::moduleEngine->module<StateMachineModule>();
+
+    module->initializeStateMachine(states, transitions, startState);
+
+    lua_settop(L, 0);
+    ghoul_assert(lua_gettop(L) == 0, "Incorrect number of items left on stack");
+    return 0;
+}
+
+int goTo(lua_State* L) {
+    ghoul::lua::checkArgumentsAndThrow(L, 1, "lua::goTo");
+    const bool isString = (lua_isstring(L, 1) != 0);
+
+    if (!isString) {
+        lua_settop(L, 0);
+        const char* msg = lua_pushfstring(
+            L,
+            "%s expected, got %s",
+            lua_typename(L, LUA_TSTRING),
+            luaL_typename(L, -1)
```

Comment:
Shouldn't this be connected to what type the actual argument was? It would make the error message clearer for users and us in the future.

Reply:
That is what this does 😄 the `luaL_typename` gets the type name of the variable at that corresponding place in the stack. However, the index is incorrect (similar to the copy paste error you noticed, where the 2s should be 1s)